### PR TITLE
Add Comment in LetSection and TypeSection

### PR DESCRIPTION
**Pull request by Unit112358**

### CODE REVIEW

File: src/il/il.nim [hunk 1]
```diff
@@ -71,18 +71,29 @@ type
         consts*: Table[string, seq[Symbol]] # deprecated
         typeOrder*: Order[Value]  # cumulative
         converters*: Table[(Value, Value), Ident]
+    DefKind* {.pure.} = enum
+        Def
+        Comment
```
Comment (chocobo333) [inline src/il/il.nim:85]:
```
Comment = object
  s: string
  isDoc: bool
```

File: src/il/il.nim [hunk 2]
```diff
@@ -71,18 +71,29 @@ type
...
     IdentDef* = ref object
-        # represents `pat: typ = default # docStr`
-        pat*: Pattern
-        typ*: Option[Expression]
-        default*: Option[Expression]
-        docStr*: Option[string]
```
Comment (chocobo333) [inline src/il/il.nim:79]: `Option[seq[Comment]]`

**Review by chocobo333 — DISMISSED** [review]
This change should have less global changes.

**Comment by Unit112358** [comment]
Revised

File: src/il/il.nim (hunk 1 above)
Comment (Unit112358) [reply src/il/il.nim:85]: ok

File: src/il/il.nim (hunk 2 above)
Comment (Unit112358) [reply src/il/il.nim:79]: ^_^

File: src/il/comment.nim [hunk 3]
```diff
@@ -0,0 +1,8 @@
+
```
Comment (chocobo333) [inline src/il/comment.nim]: The file name should be `comments.nim`

File: src/il/section.nim [hunk 4]
```diff
@@ -0,0 +1,10 @@
+
```
Comment (chocobo333) [inline src/il/section.nim]: Like comments.nim, the file name should be setions.nim

File: src/il/tos.nim [hunk 5]
```diff
@@ -178,35 +181,37 @@ proc `$`*(self: Pattern): string =
...
+        pat = $self.pat
+        typ = if self.typ.isNone: "" else: fmt": {self.typ.get}"
+        default = if self.default.isNone: "" else: fmt" = {self.default.get}"
+        docStrs = self.comments.filterIt(it.isDoc)
+        comments = self.comments.filterIt(not it.isDoc)
+    assert docStrs.len <= 1
+    let
+        docStr = if docStrs.len != 0: fmt" {docStrs[0]}" else: ""
+        comment = if comments.len != 0: "\n" & fmt"{comments}" else: ""
+    pat & typ & default & docStr & comment
```
Comment (chocobo333) [inline src/il/tos.nim]: use fmt function, like fmt"{pat}{typ}{docStr}{comment}"

File: src/il/tos.nim [hunk 6]
```diff
@@ -178,35 +181,37 @@ proc `$`*(self: Pattern): string =
...
-        pat & typ & default & docStr
-    of DefKind.Comment:
-        "#" & self.comment
+    let
+        pat = $self.pat
+        typ = if self.typ.isNone: "" else: fmt": {self.typ.get}"
+        default = if self.default.isNone: "" else: fmt" = {self.default.get}"
+        docStrs = self.comments.filterIt(it.isDoc)
+        comments = self.comments.filterIt(not it.isDoc)
+    assert docStrs.len <= 1
```
Comment (chocobo333) [inline src/il/tos.nim]: what should docStr's length be less than 1 for?

File: src/il/tos.nim [hunk 7]
```diff
@@ -178,35 +181,37 @@ proc `$`*(self: Pattern): string =
...
-            default = if self.default.isNone: "" else: fmt" = {self.default.get}"
-            docStr = if self.docStr.isNone: "" else: fmt" #{self.docStr.get}"
-        pat & typ & default & docStr
-    of DefKind.Comment:
-        "#" & self.comment
+    let
+        pat = $self.pat
+        typ = if self.typ.isNone: "" else: fmt": {self.typ.get}"
+        default = if self.default.isNone: "" else: fmt" = {self.default.get}"
+        docStrs = self.comments.filterIt(it.isDoc)
```
Comment (chocobo333) [inline src/il/tos.nim]: In this case,
```
let
  a = 3
  ## aa
  # a
  ## bb
```

File: src/parsers.nim [hunk 8]
```diff
@@ -184,7 +184,8 @@ let
 
     NewLine = preceded(+(sp0 > p"\n"), sp0) @ (it => it.len)
 
-    Comment = preceded(s"#", p".*")
+    Comment = preceded(s"#", p".*") @ (it => newComment(it))
+    DocStr = preceded(s"#", p".*") @ (it => newComment(it, true))
```
Comment (chocobo333) [inline src/parsers.nim]: Do we need to distinguish between comments and documents?
```suggestion
Comment = alt(
  preceded(s"#", p".*") @ (it => newComment(it)),
  preceded(s"#", p".*") @ (it => newComment(it, true))
```

**Review by chocobo333 — CHANGES_REQUESTED** [review]

File: src/il/comment.nim (hunk 3 above)
Comment (chocobo333) [reply src/il/comment.nim]: In my opinion, this file name implies this file has function**s** related to the `Comment` object.

File: src/il/tos.nim (hunk 5 above)
Comment (chocobo333) [reply src/il/tos.nim]: It is more intuitive to use the `fmt` function

File: src/il/comment.nim (hunk 3 above)
Comment (Unit112358) [reply src/il/comment.nim]: ok

File: src/il/section.nim (hunk 4 above)
Comment (Unit112358) [reply src/il/section.nim]: ok

File: src/il/tos.nim (hunk 5 above)
Comment (Unit112358) [reply src/il/tos.nim]: i see

File: src/il/tos.nim (hunk 6 above)
Comment (Unit112358) [reply src/il/tos.nim]: i'v thought that only the comment that is in the same line as the pattern is docStr and the others are (ordinary) comments.
so, i assumed the number of docStr would be 0 or 1.

File: src/il/tos.nim (hunk 7 above)
Comment (Unit112358) [reply src/il/tos.nim]: i think it works correctly (or my opinion that is stated in the above conversation is incorrect).

File: src/parsers.nim (hunk 8 above)
Comment (Unit112358) [reply src/parsers.nim]: il.comment has property "isDoc", so i thought we need to do.

**Comment by chocobo333** [comment]
```
## this comment is documen

**Comment by chocobo333** [comment]
```
let
  a = 3
  ## this comment is document
  # this is merely  comment
  ## this comment is also document
```
Document comments should start with "##" in anywhere

**Comment by chocobo333** [comment]
The position where the comment and the document appear is exactly the same, so
```suggestion
Comment = alt(
  preceded(s"# ", p".*") @ (it => newComment(it)),
  preceded(s"## ", p".*") @ (it => newComment(it, true))
)
```
this definition may be good

**Comment by Unit112358** [comment]
Revised

File: src/parsers.nim [hunk 9]
```diff
@@ -184,8 +184,12 @@ let
 
     NewLine = preceded(+(sp0 > p"\n"), sp0) @ (it => it.len)
 
-    Comment = preceded(s"#", p".*") @ (it => newComment(it))
-    DocStr = preceded(s"#", p".*") @ (it => newComment(it, true))
+    Comment = alt(
```
Comment (chocobo333) [inline src/parsers.nim:187]:
```
Comment = alt(
        preceded(s"##", p".*") @ (it => newComment(it, true)),
        preceded(s"#", p".*") @ (it => newComment(it))
)
```
documents start with "##"

**Review by chocobo333 — CHANGES_REQUESTED** [review]

**Review by chocobo333 — APPROVED** [review]

**Pull request merged**